### PR TITLE
[ray,rollout,trtllm] feat: Adding tensorrt_llm as new rollout engine

### PR DESCRIPTION
**Pull request by joyang-nv**
## What does this PR do?

[TensorRT-LLM](https://github.com/NVIDIA/TensorRT-LLM) has recently added [Ray orchestrator](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/ray_orchestrator) and essential features required for the RL workflow. This PR introduces TensorRT-LLM as a new rollout engine for VeRL.

VeRL currently supports several rollout modes:

- **Hybrid engine:** The training and rollout engines share the same process group. VeRL uses the `WorkerDict` class to manage multiple workers within a single process group. Communication between training and rollout workers takes place within the same process, allowing them to share the Torch GPU memory pool.
- **Colocated:** Different engines use the same set of GPUs but run in separate process groups. Currently, this mode is used only by the reward model.
- **Standalone:** Rollout engines use completely independent GPU resources.

Unlike other rollout engines, TensorRT-LLM primarily targets the *colocated* mode. However, instead of relying purely on standard colocated mode, we introduced a mixed design combining aspects of the hybrid engine and colocated mode. The design goals are:

- Clear resource separation through distinct process groups, offering maximum flexibility between training and rollout processes.
- Hybrid workers that act as proxies to LLM servers.
- Fully RESTful rollout API support through `TRTLLMHttpServer`.
- A unified framework for both asynchronous and synchronous RL workflows.

This PR aims to make the integration as minimally intrusive as possible to VeRL's infrastructure. Currently, it only invokes `RolloutReplica.init_hybrid_colocated()` when both the hybrid engine is enabled and the rollout engine is set to TensorRT-LLM.

## High Level Design
Please refer to [workers/rollout/trtllm_rollout/trtllm_async_rollout.md](https://github.com/davidmlw/verl/pull/43/changes#diff-96bab8796296991333a973a5211166f45993b13d7c533732c83bcf23c5664f39) for more details.

```mermaid
%%{init: {'theme':'base', 'themeVariables': { 'fontSize':'18px', 'edgeLabelBackground':'#eeeeee'}}}%%
flowchart TB
    space1[" "]
    style space1 fill:none,stroke:none
    
    subgraph VERL["<b>VERL Training Pipeline</b>"]
        subgraph Workers["<b>Training Workers</b>"]
            Actor["<b>Actor Worker</b>"]
            Critic["<b>Critic Worker</b>"]
            RefModel["<b>Ref Model Worker</b>"]
        end
        
        Actor -->|<b>Weight Updates<br/>IPC</b>| Rollout["<b>TensorRT-LLM Rollout</b>"]
        
        subgraph RayCluster["<b>Rollout Workers<br/>(Ray Cluster)</b>"]
            space2[" "]
            style space2 fill:none,stroke:none
            
            subgraph AsyncRollout["<b>TRTLLMAsyncRollout<br/>(per DP rank)</b>"]
                DPLeader["<b>• DP Leader coordination</b>"]
                IPCMgmt["<b>• IPC handle management</b>"]
                HTTPAdapter["<b>• HTTP adapter for server communication</b>"]
            end
            
            AsyncRollout -->|<b>HTTP/REST API</b>| HTTPServer
            
            subgraph HTTPServer["<b>TRTLLMHttpServer<br/>(Ray Actor per Replica)</b>"]
                OpenAI["<b>• OpenAI Server wrapper</b>"]
                EngMgmt["<b>• AsyncLLM engine management</b>"]
                MemMgmt["<b>• Memory management (resume/release)</b>"]
            end
            
            HTTPServer --> AsyncLLM
            
            subgraph AsyncLLM["<b>TensorRT-LLM<br/>AsyncLLM Engine</b>"]
                GPUWorkers["<b>• GPU workers (Tensor Parallel)</b>"]
                KVCache["<b>• KV Cache management</b>"]
                CUDAGraph["<b>• CUDA Graph optimization</b>"]
            end
        end
    end
    
    space1 ~~~ VERL
    
    style VERL fill:#e1f5ff
    style RayCluster fill:#fff4e6
    style AsyncRollout fill:#f3e5f5
    style HTTPServer fill:#e8f5e9
    style AsyncLLM fill:#fce4ec
```

## Experiments results:
Setup: single node with H100 * 8/slurm env.

1. FSDP/GRPO: Qwen2-7B (TP1 * 8 on 8 GPUs, launching cmd `bash examples/grpo_trainer/run_qwen2-7b_math_trtllm.sh 1`)
    * Convergence:    

   * Validation:

2. FSDP/GRPO: Qwen2-7B (TP4 * 2 on 8 GPUs, launching cmd `bash examples/grpo_trainer/run_qwen2-7b_math_trtllm.sh 4`)

    * Convergence:

    * Validation:    

3. Megatron/GRPO: Qwen2-7B (TP1 * 8 on 8 GPUs, launching cmd `bash examples/grpo_trainer/run_qwen2-7b_math_megatron_trtllm.sh 1`)

    * Convergence: 
      image

    * Validation:   
      image

4. Megatron/GRPO: Qwen2-7B (TP2 * 2 on 8 GPUs, launching cmd `bash examples/grpo_trainer/run_qwen2-7b_math_megatron_trtllm.sh 4`)
    
    * Convergence:
      image

    * Validation:
      image

## Special notes for using VeRL with TensorRT-LLM:
1. All RL required APIs for VeRL were implemented within [TensorRT-LLM 1.2.0rc6](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/tensorrt-llm/containers/release?version=1.2.0rc6). To install VeRL with TensorRT-LLM, please use command `pip install -e ".[trtllm]" --extra-index-url https://pypi.nvidia.com/`.
2. All verification of integration work was primarily done in Slurm environment.
3. The current design requires `export RAY_EXPERIMENTAL_NOSET_CUDA_VISIBLE_DEVICES=1` and the following env settings before launching the Ray cluster. While these have been included in any example scripts or tests added, we will work toward removing such dependencies to improve the user experience in the near future. 
```
# Clean all slurm / MPI / PMIx env to avoid pmix mismatch error
for v in $(env | awk -F= '/^(PMI|PMIX|MPI|OMPI|SLURM)_/{print $1}'); do
    unset "$v"
done
# Force UCX to use only eth0; otherwise, it will attempt to use all available devices and raise warnings if any issues occur.
export TRTLLM_UCX_INTERFACE=eth0
```

## Outstanding issues for this MR:
1. WIP on passing CI tests 

## Upcoming works (in separate MRs)
1. Further performance optimization
3. Multi-node testing and functionality will be delivered in the near future. 
7. The current MR focuses on and was validated wtih Qwen model variants. We'll work on validations and optimizations for MoE models as the next step.

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: docker/Dockerfile.stable.trtllm [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    pip3 install --no-cache-dir nvtx matplotlib liger_kernel && \
+    pip install --no-cache-dir -U git+https://github.com/ISEEKYAN/mbridge.git && \
+    pip install --no-deps --no-cache-dir git+https://github.com/NVIDIA/Megatron-LM.git@core_v0.14.0rc7
+
+
+# ==============================================================================
+# Install verl dependencies
+# ==============================================================================
+RUN pip install git+https://github.com/volcengine/verl.git@v0.6.0
+RUN pip uninstall -y verl
```
[the remark this answers is not in the document]
Comment (Superjomn) [reply docker/Dockerfile.stable.trtllm:65]: This snippet is borrowed from [Dockerfile.stable.vllm](https://github.com/volcengine/verl/blob/main/docker/Dockerfile.stable.vllm#L112), and we prefer to keep it in this PR.

File: verl/trainer/config/rollout/rollout.yaml [hunk 2]
```diff
@@ -58,6 +58,9 @@ expert_parallel_size: 1
 # PP size for rollout.
 pipeline_model_parallel_size: 1
 
+# max batch size for rollout
+max_batch_size: 256
```
Comment (wuxibin89) [inline verl/trainer/config/rollout/rollout.yaml]: Duplicated with `max_num_seqs`

File: verl/workers/rollout/trtllm_rollout/trtllm_rollout.py [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+        Args:
+            weights: A dictionary that maps the device uuid of the weight handles.
+
+        Returns:
+            Dict[str, Any]: Server response containing update status
+        """
+        return await self._make_async_request("update_weights", {"weights": weights})
+
+
+class TRTLLMAsyncRollout(BaseRollout):
```
Comment (wuxibin89) [inline verl/workers/rollout/trtllm_rollout/trtllm_rollout.py]: Naming: unify all backends to `ServerAdapter`

File: verl/workers/rollout/trtllm_rollout/trtllm_rollout.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+        for name, param in weights:
+            size_in_bytes = param.element_size() * param.numel()
+            if size_in_bytes > cur_available_bytes:
+                await flush()
+
+            assert cur_available_bytes >= size_in_bytes, (
+                f"cur_available_bytes: {cur_available_bytes:,} size_in_bytes: {size_in_bytes:,} name: {name}"
+            )
+            cur_available_bytes -= size_in_bytes
+            handle = reduce_tensor(param.detach())
```
Comment (wuxibin89) [inline verl/workers/rollout/trtllm_rollout/trtllm_rollout.py:441]: Cuda ipc creation is pretty heavy, we should create cuda ipc once and reuse it for all bucket.
https://github.com/volcengine/verl/pull/4280/files#diff-57402386e9d4649fe5930175ba001192e250e2d21a21ecc5000d556074df5113R154-R183

File: verl/single_controller/ray/base.py [hunk 5]
```diff
@@ -508,7 +509,10 @@ def _init_with_subresource_pool(self, resource_pool, ray_cls_with_init, bin_pack
 
         rank = -1
         local_world_size = resource_pool.store[0]
-        self._get_master_addr_port(pgs[0])
+        self._get_master_addr_port(
```
Comment (wuxibin89) [inline verl/single_controller/ray/base.py:514]: @yyDing1 This is a hidden bug from resource pool split?

File: verl/workers/rollout/replica.py [hunk 6]
```diff
@@ -125,6 +126,22 @@ async def init_hybrid(self, worker_group: RayWorkerGroup):
         ]
         await self.launch_servers()
 
+    async def init_hybrid_colocated(self, worker_group: RayWorkerGroup, resource_pool: RayResourcePool):
```
Comment (wuxibin89) [inline verl/workers/rollout/replica.py:131]: I think we can modify `init_hybrid` directly to extract resource pool from worker_group.
```
    async def init_hybrid(self, worker_group: RayWorkerGroup):
        self.rollout_mode = RolloutMode.HYBRID
        self.workers = worker_group.workers[
            self.world_size * self.replica_rank : self.world_size * (self.replica_rank + 1)
        ]
        self.resource_pool = worker_group.resource_pool
        await self.launch_servers()
```

File: verl/workers/config/rollout.py [hunk 7]
```diff
@@ -189,6 +190,7 @@ class RolloutConfig(BaseConfig):
     custom: Optional[dict] = None
 
     update_weights_bucket_megabytes: int = 512
+    refit_ipc_memory_ratio: float = 0.5
```
Comment (wuxibin89) [inline verl/workers/config/rollout.py]: Can we reuse `update_weights_bucket_megabytes`? It is used for both vllm and sglang.

File: verl/workers/rollout/replica.py [hunk 8]
```diff
@@ -196,6 +216,9 @@ def server_handle(self) -> ActorHandle:
         """Get rollout server handle for Token-in-token-out generation."""
         return self._server_handle
 
+    def rollout_worker_use_gpu(self) -> bool:
```
Comment (wuxibin89) [inline verl/workers/rollout/replica.py:221]: This is a bit redundant, rollout should always use GPU

File: verl/experimental/agent_loop/agent_loop.py [hunk 9]
```diff
@@ -891,10 +898,19 @@ def _initialize_llm_servers(self):
...
             for replica_rank in range(num_replicas)
         ]
-        if self.worker_group:
+
+        if self.worker_group and rollout_config.name != "trtllm":
             self._run_all([server.init_hybrid(self.worker_group) for server in self.rollout_replicas])
+        elif self.worker_group and rollout_config.name == "trtllm":
+            self._run_all(
+                [
+                    server.init_hybrid_colocated(self.worker_group, rollout_resource_pool)
```
Comment (wuxibin89) [inline verl/experimental/agent_loop/agent_loop.py:908]: Please unify to `init_hybrid`

File: verl/workers/rollout/replica.py (hunk 8 above)
Comment (joyang-nv) [reply verl/workers/rollout/replica.py:221]: TRTLLM design for `init_colocated()` mode will NOT use GPU since it is just proxy to TRTLLM server. :)

File: verl/single_controller/ray/base.py (hunk 5 above)
Comment (yyDing1) [reply verl/single_controller/ray/base.py:514]: Yes, but this env variable seems not to be used in the previous rollout engine init.

File: verl/workers/rollout/trtllm_rollout/trtllm_rollout.py (hunk 4 above)
Comment (Superjomn) [reply verl/workers/rollout/trtllm_rollout/trtllm_rollout.py:441]: Good suggestion. Single ipc handle with a buffer could be ideal, but it needs to introduce another copy from the weight tensors to the buffer, right? 
Considering that the update_weights is implemented in the [trtllm](https://github.com/shuyixiong/TensorRT-LLM/blob/166c8352ef3ef0f43d319a6e9b2da0aa1a2a83ec/tensorrt_llm/llmapi/rlhf_utils.py#L36) repo, and updating it may need more time (for an RC release), can we optimize this in a separate PR?

Comment (wuxibin89) [reply verl/workers/rollout/trtllm_rollout/trtllm_rollout.py:441]: Sure

File: verl/single_controller/ray/base.py (hunk 5 above)
Comment (wuxibin89) [reply verl/single_controller/ray/base.py:514]: But sub resource pool may be also used for model engine, e.g teacher model for online policy distilation.

File: verl/workers/rollout/trtllm_rollout/trtllm_async_server.py [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+            bundle_indices.append(pg_bundle_indices)
+            left_bundle_count -= left_bundle_count_in_pg
+            local_bundle_index = 0
+
+        assert left_bundle_count == 0, "all bundle indices should be assigned"
+
+        return pgs, bundle_indices
+
+    async def launch_servers(self):
+        assert self.nnodes == 1, "TRTLLMReplica doesn't support multiple nodes for single replica yet."
```
Comment (wuxibin89) [inline verl/workers/rollout/trtllm_rollout/trtllm_async_server.py:285]: Is this a restriction of TRT-LLM or verl integration?

Comment (joyang-nv) [reply verl/workers/rollout/trtllm_rollout/trtllm_async_server.py:285]: Yeap, at this moment, TRTLLM only support replica within single node. Our design is for unifying multi-node but we had not verified it yet and just keep assertion till we verified it.
